### PR TITLE
util: support spill offset into disk when spilling

### PR DESCRIPTION
**Pull request by wshwsh12**
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: ref https://github.com/pingcap/tidb/issues/33877

Problem Summary:

### What is changed and how it works?

Support spilling offsets of each row in ListInDisk, which reduces the memory usage for spilling.

Benchmark:
```
Master:
BenchmarkListInDiskAdd
BenchmarkListInDiskAdd-8   	 2148649	       526.5 ns/op
BenchmarkListInDiskGetRow
BenchmarkListInDiskGetRow-8   	  403568	      2814 ns/op

This PR:
BenchmarkListInDiskAdd
BenchmarkListInDiskAdd-8   	 1991194	       576.6 ns/op
BenchmarkListInDiskGetRow
BenchmarkListInDiskGetRow-8   	  318237	      3696 ns/op
```

Heap Pprof: Sort 1e8 rows and spill. We can see the memory usage of chunkInDisk.WriteTo disappeared
```
Master:
Showing nodes accounting for 1760.58MB, 98.76% of 1782.60MB total
Dropped 212 nodes (cum <= 8.91MB)
Showing top 10 nodes out of 34
      flat  flat%   sum%        cum   cum%
  762.96MB 42.80% 42.80%   762.96MB 42.80%  github.com/pingcap/tidb/util/chunk.(*SortedRowContainer).Sort
  733.70MB 41.16% 83.96%   733.70MB 41.16%  github.com/pingcap/tidb/util/chunk.(*chunkInDisk).WriteTo
  204.52MB 11.47% 95.43%   204.52MB 11.47%  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse

This PR:
Showing nodes accounting for 1041.51MB, 97.18% of 1071.69MB total
Dropped 238 nodes (cum <= 5.36MB)
Showing top 20 nodes out of 33
      flat  flat%   sum%        cum   cum%
  762.95MB 71.19% 71.19%   762.95MB 71.19%  github.com/pingcap/tidb/util/chunk.(*SortedRowContainer).Sort
  204.52MB 19.08% 90.28%   204.52MB 19.08%  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse
   74.03MB  6.91% 97.18%    74.03MB  6.91%  google.golang.org/grpc.(*parser).recvMsg
         0     0% 97.18%    74.03MB  6.91%  github.com/pingcap/kvproto/pkg/tikvpb.(*tikvBatchCommandsClient).Recv
```

Performance Test: Sort 1e8 rows and set memory quota to 1GB, tidb will spill data to disk. This pr will use more disk space.
```
Master:
[tidb]> desc analyze  select * from v10kw big_table order by big_table.a;
+------------------------------+--------------+-----------+-----------+------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------+---------+
| id                           | estRows      | actRows   | task      | access object    | execution info                                                                                                                                                                                                                                                                                                                      | operator info             | memory  | disk    |
+------------------------------+--------------+-----------+-----------+------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------+---------+
| Sort_9                       | 100000000.00 | 100000000 | root      |                  | time:6m5.5s, loops:97658                                                                                                                                                                                                                                                                                                            | test.base_table.a         | 1.24 GB | 2.88 GB |
| └─Limit_12                   | 100000000.00 | 100000000 | root      |                  | time:23.9s, loops:97720                                                                                                                                                                                                                                                                                                             | offset:0, count:100000000 | N/A     | N/A     |
|   └─TableReader_16           | 100000000.00 | 100000554 | root      |                  | time:23.9s, loops:97719, cop_task: {num: 107, max: 1.47s, min: 198.9ms, avg: 682.2ms, p95: 1.26s, max_proc_keys: 1436388, p95_proc_keys: 1340635, tot_proc: 1m2.7s, tot_wait: 1.06s, rpc_num: 107, rpc_time: 1m13s, copr_cache_hit_ratio: 0.00}                                                                                     | data:Limit_15             | 93.9 MB | N/A     |
|     └─Limit_15               | 100000000.00 | 100248665 | cop[tikv] |                  | tikv_task:{proc max:1.15s, min:173ms, p80:763ms, p95:872ms, iters:98390, tasks:107}, scan_detail: {total_process_keys: 100248665, total_process_keys_size: 3709200605, total_keys: 100248772, rocksdb: {delete_skipped_count: 0, key_skipped_count: 100248665, block: {cache_hit_count: 76095, read_count: 0, read_byte: 0 Bytes}}} | offset:0, count:100000000 | N/A     | N/A     |
|       └─TableFullScan_14     | 100000000.00 | 100248665 | cop[tikv] | table:base_table | tikv_task:{proc max:1.15s, min:173ms, p80:763ms, p95:872ms, iters:98390, tasks:107}                                                                                                                                                                                                                                                 | keep order:false          | N/A     | N/A     |
+------------------------------+--------------+-----------+-----------+------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------+---------+
5 rows in set (6 min 5.717 sec)

This PR:
[tidb]> desc analyze  select * from v10kw big_table order by big_table.a;
+------------------------------+--------------+-----------+-----------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------+---------+
| id                           | estRows      | actRows   | task      | access object    | execution info                                                                                                                                                                                                                                                                                                                         | operator info             | memory  | disk    |
+------------------------------+--------------+-----------+-----------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------+---------+
| Sort_9                       | 100000000.00 | 100000000 | root      |                  | time:7m36.8s, loops:97658                                                                                                                                                                                                                                                                                                              | test.base_table.a         | 1.20 GB | 3.55 GB |
| └─Limit_12                   | 100000000.00 | 100000000 | root      |                  | time:26.9s, loops:97720                                                                                                                                                                                                                                                                                                                | offset:0, count:100000000 | N/A     | N/A     |
|   └─TableReader_16           | 100000000.00 | 100000554 | root      |                  | time:26.9s, loops:97719, cop_task: {num: 107, max: 2.18s, min: 196.4ms, avg: 947.8ms, p95: 2.09s, max_proc_keys: 1436388, p95_proc_keys: 1340635, tot_proc: 1m3.5s, tot_wait: 220ms, rpc_num: 107, rpc_time: 1m41.4s, copr_cache_hit_ratio: 0.00}                                                                                      | data:Limit_15             | 97.7 MB | N/A     |
|     └─Limit_15               | 100000000.00 | 100248665 | cop[tikv] |                  | tikv_task:{proc max:1.26s, min:168ms, p80:757ms, p95:974ms, iters:98390, tasks:107}, scan_detail: {total_process_keys: 100248665, total_process_keys_size: 3709200605, total_keys: 100248772, rocksdb: {delete_skipped_count: 0, key_skipped_count: 100248665, block: {cache_hit_count: 200, read_count: 75895, read_byte: 122.1 MB}}} | offset:0, count:100000000 | N/A     | N/A     |
|       └─TableFullScan_14     | 100000000.00 | 100248665 | cop[tikv] | table:base_table | tikv_task:{proc max:1.26s, min:168ms, p80:757ms, p95:974ms, iters:98390, tasks:107}                                                                                                                                                                                                                                                    | keep order:false          | N/A     | N/A     |
+------------------------------+--------------+-----------+-----------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------+---------+---------+
5 rows in set (7 min 37.166 sec)
```


### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: util/chunk/disk_test.go [hunk 1]
```diff
@@ -145,25 +143,47 @@ type listInDiskWriteDisk struct {
 	ListInDisk
 }
 
+func (l *tempFileWithIOWrapper) flushForTest() (err error) {
```
Comment (XuHuaiyu) [inline util/chunk/disk_test.go:146]: It seems `flush` is not an appropriate name for now?

File: util/chunk/disk.go [hunk 2]
```diff
@@ -33,16 +33,17 @@ import (
 type ListInDisk struct {
 	fieldTypes         []*types.FieldType
 	numRowForEachChunk []int
```
Comment (XuHuaiyu) [inline util/chunk/disk.go:35]: numRowsOfEachChunk
rowNumOfEachChunkFirstRow
totalNumRows

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**